### ISSUE TITLE
clock epochs 

### ISSUE THREAD

**Issue by JaapAap** [HowardHinnant/date#129]
I noticed that clock epochs don't seem to be stored anywhere but rather live in the code that does conversion between different clocks. As a result, tz.h contains the code "sys_days{year{1970}/jan/1}" in 6 locations. While there might be good reasons for doing so, it would be nice if this would be stored in a single location that is also programmatically accessible.  

I want to introduce a few more user-defined 'clocks' and it would be nice to have access to the epoch rather than having to repeat it again and again. On a related note: it would also be nice to have a generic clock conversion routine that uses this information (but would in addition have to get access to the offset between the epochs in the different time scales that is similarly defined twice in the code for each pair of 'clocks' - note sure where that should be defined though).

**Comment by HowardHinnant** [comment]
Here is a generic function for getting a clock's epoch in terms of `sys_time`:
````
template <class Clock>
auto
get_epoch()
{
    return date::to_sys_time(typename Clock::time_point{});
}
````
Here is example code doing this for `tai_clock`.
````
int
main()
{
    using namespace date;
    std::cout << get_epoch<tai_clock>() << '\n';
}
````
which will output:
````
1957-12-31 23:59:50.000000
````

**Comment by JaapAap** [comment]
Dear Howard,

Thanks for that. I have to admit I was more thinking upon the lines of implementing this:

template <class to_type, class from_type>
to_type
clock_cast(const from_type& t)
{
	return to_type{ t.time_since_epoch() } + (from_type::clock::epoch() - to_type::clock::epoch());
}

by defining static member functions as part of the clock definitions implemented as:

std::chrono::system_clock::time_point 
tai_clock::epoch() NOEXCEPT
{
	using namespace std::chrono;
	return sys_days{ year{ 1958 } / jan / 1 } - seconds{ 10 };
}:

I'd be happy to share my implementation (which can prob. be improved upon, this is rather a proof-of-concept)...

**Comment by HowardHinnant** [comment]
There be dragons here:

The difference between `system_clock`'s epoch, and the other three clock's epoch, is not a constant.  It varies with time.  For example the epoch difference between `utc_clock` and `system_clock` was 0s in 1970 and is today 27s.

If one calls `tai_clock::epoch()` it will be surprising to some if the `time_point` is not returned as a `tai_clock::time_point`, and surprising to others if it is not returned as a `system_clock::time_point`.  Thus I encourage you to write whatever makes sense for you.

Note that if you _do_ convert everything to (or through) `sys_time`, you will loose leap second accuracy when you compute the epoch offset.

You can always depend on the following invariants:

1.  `Clock::time_point{}` will give you the `Clock`'s epoch in terms of `Clock::time_point`.
2.  `to_Clock2_time(Clock1::time_point)` will _exactly_ convert from `Clock1` to `Clock2` as long as neither `Clock` is `std::chrono::system_clock`.  If one of them _is_ `std::chrono::system_clock` then you get an approximate mapping  that  does not count leap seconds.

Here is an equivalent of your code above:
````
template <class to_type, class from_type>
to_type
clock_cast(const from_type& t)
{
    return to_type{ t.time_since_epoch() } + (to_sys_time(from_type{}) - to_sys_time(to_type{}));
}
````
But it has a couple of problems.  The first is that it gives the wrong answer when (for example) you convert from `gps_seconds` to `tai_seconds`.  If you enter `gps_now()` it gives a difference of 10s, but the correct difference is 19s.  This is caused by the trip through `sys_time`.  If this is changed to go through `utc_time` instead:

````
template <class to_type, class from_type>
to_type
clock_cast(const from_type& t)
{
    return to_type{ t.time_since_epoch() } + (to_utc_time(from_type{}) - to_utc_time(to_type{}));
}
````
Now you get the correct result.  However this still has a bug that it is missing an identity transformation from `utc_time` to `utc_time`.  That transformation would be easy enough to add, and perhaps I'll do that (need to give it more thought).

**Comment by HowardHinnant** [comment]
Oh, the above still won't work if the source or destination is `sys_time`, again because of the shifting epoch difference.  One has to take the difference between the times at the current `time_point` instead of at the epoch.

It's confusing stuff.  Here be dragons.

**Comment by JaapAap** [comment]
I am confused indeed - i have to admit that this is not my normal territory. I seems that the term 'epoch' already is confusing in itself. You write "The difference between  system_clock 's epoch, and the other three clock's epoch, is not a constant". I was trying to refer to the 'reference epoch', that is, using epoch as in the 'time_since_epoch()' call as the start of time for a particular time scale. Maybe I should have named it ref_epoch() or even ref_epoch_in_sys_time() to avoid (some) confusion. Why wouldn't it be sufficient to use translate the reference epochs (which for GPS includes 9 leap seconds wrt sys_time) and durations since the reference epoch for each time scale? As far as I can see it gives the exact same code that the current  to_Clock2_time functions do?

**Comment by HowardHinnant** [comment]
It's easier to visualize if we talk about `utc_clock` and `system_clock`.  These both use an epoch of 1970-01-01 00:00:00 UTC.  So on that epoch date/time,  the difference between their epochs is 0s, right?  This can be confirmed with the following program:
````
#include "chrono_io.h"
#include "tz.h"
#include <iostream>

int
main()
{
    using namespace date;
    using namespace std;
    sys_seconds s = sys_days{1970_y/jan/1};
    utc_seconds u = to_utc_time(s);
    cout << u.time_since_epoch() - s.time_since_epoch() << '\n';
}
````
which outputs:
````
0s
````
Now if we do nothing but change to today's date:
````
    sys_seconds s = sys_days{2017_y/feb/13};
````
the output becomes:
````
27s
````
This is because `system_clock` does not count leap seconds, and `utc_clock` does.  And we get similar leap second issues between `system_clock` and the other two clocks.  That being said, if we just look at clocks besides `system_clock`, then the epoch difference is constant.  Here for example is the difference between `tai_clock` and `utc_clock` (both of these clocks count leap seconds):
````
#include "chrono_io.h"
#include "tz.h"
#include <iostream>

int
main()
{
    using namespace date;
    using namespace std;
    utc_seconds u = to_utc_time(sys_days{1970_y/jan/1});
    tai_seconds t = to_tai_time(u);
    cout << t.time_since_epoch() - u.time_since_epoch() << '\n';
}
````
This outputs:
````
378691210s
````
If you change the date to today, the output is the same.  So your algorithm will work fine translating between `utc_clock` and `tai_clock`, but not between `utc_clock` and `system_clock`.

As `system_clock` is the only clock among these four that doesn't count leap seconds, it is the only clock that causes trouble if used with your algorithm.

**Comment by JaapAap** [comment]
First of all thanks for your extensive reply, which is quite instructive. It seems I am still missing some understanding though . Let's look at your code translating GPS to UTC:

utc_time<...>
to_utc_time(const gps_time<Duration>& t)
{...   
  return utc_time<duration>{t.time_since_epoch()} +
            (sys_days{year{1980}/jan/sun[1]} - sys_days{year{1970}/jan/1} + seconds{9});
}

I see the reference epoch of UTC expressed in system time (sys_days{year{1970}/jan/1}) is subtracted from the reference epoch of GPS expressed in system  time (sys_days{year{1980}/jan/sun[1]} + seconds
{9}). My function does the exact same thing, but it uses the reference epochs returned by the 'clock' class (defined as above) instead of using the to allow their re-use throughout all potential timescale conversions). So in my understanding the function replicates all the time1_to_time2 functions perfectly (and I get the same results from each). I don't believe there is a unit test yet for these transformation that I could replicate?

Admittedly I might still be missing something; one thing I certainly don't grasp is whether you would indeed expect your get_epoch<gps_clock>() to output a system time of  "1980-06-01 00:00:00.000000" - as in 'my' implementation, it would be returned as "1980-06-01 00:00:09.0000" instead:

class gps_clock
{
public:
...
static std::chrono::system_clock::time_point epoch() NOEXCEPT;
...
};

std::chrono::system_clock::time_point
gps_clock::epoch() 
{ 
	using namespace std::chrono;
	return sys_days{ year{ 1980 } / jan / sun[1] } + seconds{ 9 };
}

**Comment by HowardHinnant** [comment]
Yes, your algorithm works when translating between `gps_time` and `utc_time`, because both of these clocks count leap seconds.  Thus the difference between their epochs is constant as time progresses.

`sys_time` is used as an implementation detail in expressing this epoch difference as an implementation convenience.  `gps_time` begins on the first Sunday in January of 1980 at 00:00:00 UTC.  9 leap seconds were added to UTC between 1970 and 1980, so that's where the 9s comes from (https://en.wikipedia.org/wiki/Leap_second).  This could have been coded as 315964809s, which is 86400 seconds for each of the 3657 days, plus an extra 9s for the leap seconds between these two times.

But the difference between `gps_time` and `sys_time` (as opposed to `utc_time`) grows by 1s every time a leap second occurs after 1980.

Here's the code to convert `gps_time` to `sys_time`, which traffics through `gps_time` to `utc_time` (adds 315964809s), and then calls https://github.com/HowardHinnant/date/blob/master/tz.h#L1078-L1100 which goes to the tzdatabase to find out how many leap seconds it needs subtract to convert `utc_time` to `sys_time` (which is a function of the `utc_time` `time_point`).

Now as to output (streaming):  `sys_time` and `utc_time` always output the same date, except during a leap second insertion.  During a leap second insertion, `sys_time` stalls at the last representable instant for a time as demo'ed in the example under https://howardhinnant.github.io/date/tz.html#gps_clock, and then `sys_time` and `utc_time` sync back up after the leap second.  But `gps_time` (and `tai_time`) have different behavior during a leap second.  They don't freeze up like `sys_time`, instead they keep ticking, but don't have a 61s minute.  They report a time 1s further in the future than `utc_time` every time a leap second occurs.

Here's a demo:
````
#include "tz.h"
#include <iostream>

int
main()
{
    using namespace date;
    using namespace std;
    using namespace std::chrono;
    utc_seconds u = to_utc_time(sys_days{1980_y/jan/sun[1]});
    gps_seconds g = to_gps_time(u);
    cout << u << " UTC == " << g << " GPS\n";
    u = to_utc_time(sys_days{2017_y/feb/13});
    g = to_gps_time(u);
    cout << u << " UTC == " << g << " GPS\n";
}
````
This outputs:
````
1980-01-06 00:00:00 UTC == 1980-01-06 00:00:00 GPS
2017-02-13 00:00:00 UTC == 2017-02-13 00:00:18 GPS
````
which says that at the `gps_clock` epoch, GPS and UTC were sync'd in how they reported "human time".  But now 37 years later, `gps_clock` reports a calendar time 18s ahead of `utc_clock` because although `gps_clock` _does_ measure leap seconds, it does not have any 61s minutes.  If you check with (https://en.wikipedia.org/wiki/Leap_second) you'll count 18 leap seconds inserted since 1980.

Now this isn't to say that the _epochs_ between `utc_clock` and `gps_clock` have changed, they haven't.  That difference is a steady 315964809s.  But how those count of seconds gets reported out as a _calendar time_ does change with time because one clock has 61s minutes every once in awhile and the other doesn't.

And if this hasn't completely confused you, you're a better man than I.  **I** find it confusing, and had to study it closely, code it up before forgetting my understanding, and now rely on my code to explain it to me again. ;-)  But it is all consistent with independently developed libraries and such as this one:  http://www.leapsecond.com/java/gpsclock.htm .

**Comment by JaapAap** [comment]
Thanks for that – indeed, still somewhat confused, but slowly getting there I believe. My understanding is (and was) that times are stored as a reference epoch + time duration and that the output streaming computes what a clock would show for a certain epoch – that is where the leap seconds play their part. [I am not sure why you state that GPS clocks count leap seconds?]. To me this still implies that all stored values can still be converted without having to worry about the leap seconds part – which only come into play when translating the stored time to the value to display.

 

Indeed, I misunderstood the behaviour of the system clock. You say that “sys_time and utc_time sync back up after the leap second.”, which means that (unlike what I understood before) that the system time scale does contain leap seconds (but with a slight twist during leap second insertion in that it displays a ‘two-second-long second rather than an additional 61-th second). I apologize for missing your example output which is making this all very clear. Confusion here was mainly caused by the many statements I read that the system clock is free of leap seconds – which clearly isn’t the case; it rather sneaks them in without explicitly displaying them. Hence, my terminology I used was incorrect and the references stored are not in system time, but in some other, leap-second free timescale (which is the reason why they work).

**Comment by HowardHinnant** [comment]
I'm afraid that my explanations have only confused you more, and clarified nothing.  Please forget everything I've said, and let's start over.  I strongly encourage you to play with the sample code as you read the explanations.  Print things out and see if they match your current understanding.

Each clock has two associated "data structures".

1.  A `chrono::time_point`.
2. A text representation when streamed out showing year, month, day, hour, minute, second, etc.

In general they can all be different among the four clocks, even for the same instant in time.

`system_clock` counts non-leap seconds as described here:  https://en.wikipedia.org/wiki/Unix_time .  That is, the duration in `system_clock::time_point` is just a count of time (seconds for example), _excluding_ leap seconds.  When this count is streamed out, it will show the same field values as UTC for a UTC `time_point` representing the same instant in time.  Since `system_clock` can not represent time during a leap second insertion, there is no way for it to print out what UTC prints out during a leap second insertion.

`utc_clock` counts all seconds since 1970 (leap and non-leap).  Thus the number of seconds represented in the `time_point` will be 1 greater than that of `system_clock::time_point` for every leap second that occurs prior to that`time_point`.  As noted above, for times not during a leap second insertion, when printed out it will have the same field values as a `system_clock::time_point` representing the same instant in time (even though the two `time_point`s will be holding a different number of seconds internally).  _During_ a leap second insertion, a `utc_clock::time_point` that is streamed out will have `60` in the seconds field.  If you convert such a `time_point` back to `system_clock::time_point` it will be the last representable instant prior to the leap second insertion.

`gps_clock` counts al seconds (leap and non-leap) since 1980-01-06 00:00:00 UTC.  It holds this count in the `time_point`.  When streamed out, this `time_point` does not always have the same field values as `utc_clock`.  For one, it will never show 60 in the seconds field.  Because of this, the apparent time it displays when printed gets further out of sync with UTC by 1s, every time there is a leap second.

Sample code to explore with:
````
#include "chrono_io.h"
#include "tz.h"
#include <iostream>

int
main()
{
    using namespace date;
    using namespace std;
    using namespace std::chrono;
    sys_seconds s = sys_days{2017_y/feb/14} + 9h + 3min + 2s;
    utc_seconds u = to_utc_time(s);
    gps_seconds g = to_gps_time(s);
    cout << "SYS: " << s.time_since_epoch() << " : " << s << '\n';
    cout << "UTC: " << u.time_since_epoch() << " : " << u << '\n';
    cout << "GPS: " << g.time_since_epoch() << " : " << g << '\n';
}
````
Output:
````
SYS: 1487062982s : 2017-02-14 09:03:02
UTC: 1487063009s : 2017-02-14 09:03:02
GPS: 1171098200s : 2017-02-14 09:03:20
````
This is the same instant in time, being represented by 3 different clocks, showing both the duration internal to the `time_point`, and the text representation when printed out.  Note that the `utc_seconds` holds 27 more seconds than does the `sys_seconds`.  And yet they output the same fields.  The difference in seconds between these two `time_points` will always be the number of leap second insertions between the time_point and 1970.

The GPS `time_point` will always hold 315964809s less than the corresponding UTC `time_point`, even during a leap second insertion.  The difference between a SYS `time_point` and a GPS `time_point` depends on how many leap seconds have been inserted after 1980-01-06 00:00:00 UTC and the `time_point`.  The text fields output by a GPS `time_point` will never contain 60 in the seconds field, and will be ahead of UTC by 1s for every leap second inserted since 1980.

**Comment by JaapAap** [comment]
Thanks Howard, I don’t belief I got confused more, but I did need another lesson in clock-ese.  I now finally understand how the word ‘count’ is being used in the context of leap seconds. In my own words, it means that all clocks ‘progress’ during a leap second except the system clock, hence ‘leap seconds’ are counted by UTC and GPS clocks alike. The system clock is special in the sense that it does skip leap seconds and its ‘counts’  get behind the counts of the other clocks. While for the GPS time scale all seconds are the same, the UTC time scale labels some of the seconds as being leap seconds which impacts what you call the textual representation (to keep it lined up with UT1). That is, for UTC, when looking at the displayed time, a leap second can be seen as a ‘specially labelled second’ which causes the textual representation to ‘get behind’ the one of GPS.

The main cause of confusion is in the area of semantics. UTC ‘gets behind GPS’ in its textual representation while representing the same epoch, while UTC gets ahead of the system time in terms of its internal representation while still having the same textual representation (except during the leap second insertion). This requires one to formulate really carefully: while the GPS timescale does not contain leap seconds (in its representation), it does ‘count’ seconds that are called ‘leap seconds’ within the UTC timescale (while system time does not). Thanks again for clarifying!

Attached is the solution I finally settled for to cast the clocks - now including a specialization for the system clock. All transformation info now lives in the to_utc_time methods and can be re-used for the additional clocks that I need in my application. This of course requires the to_utc_time method to be overloaded for utc_times as well.

[clock_cast.txt](https://github.com/HowardHinnant/date/files/775174/clock_cast.txt)
